lib/puppet/provider/exec/powershell.rb

```diff
@@ -25,17 +28,69 @@
+
+Unfortunately this library prevents the PowerShell module from using a shared
+PowerShell process to dramatically improve the performance of resource
+application.
+
+To enable these improvements, it is suggested to upgrade to any x64 version of
+Puppet (including 3.x), or to a Puppet version newer than 3.x.
+    UPGRADE
+  end
+
+  def self.win32console_enabled?
+    @enabled ||= defined?(Win32::Console) && Win32::Console.class == Class
+  end
+
+  if self.win32console_enabled?
+    upgrade_message
+  end
+
+  def self.powershell_args
+    ps_args = ['-NoProfile', '-NonInteractive', '-Sta', '-NoLogo', '-ExecutionPolicy', 'Bypass', '-Command']
+    ps_args << '-' if !self.win32console_enabled?
+    ps_args
+  end
+
+  def ps_manager
+    PuppetX::PowerShell::PowerShellManager.instance("#{command(:powershell)} #{self.class.powershell_args.join(' ')}")
+  end
+
   def run(command, check = false)
-    write_script(command) do |native_path|
-      # Ideally, we could keep a handle open on the temp file in this
-      # process (to prevent TOCTOU attacks), and execute powershell
-      # with -File <path>. But powershell complains that it can't open
-      # the file for exclusive access. If we close the handle, then an
-      # attacker could modify the file before we invoke powershell. So
-      # we redirect powershell's stdin to read from the file. Current
-      # versions of Windows use per-user temp directories with strong
-      # permissions, but I'd rather not make (poor) assumptions.
-      return super("cmd.exe /c \"\"#{native_path(command(:powershell))}\" #{args} -Command - < \"#{native_path}\"\"", check)
+    if self.class.win32console_enabled?
+      write_script(command) do |native_path|
+        # Ideally, we could keep a handle open on the temp file in this
+        # process (to prevent TOCTOU attacks), and execute powershell
+        # with -File <path>. But powershell complains that it can't open
+        # the file for exclusive access. If we close the handle, then an
+        # attacker could modify the file before we invoke powershell. So
+        # we redirect powershell's stdin to read from the file. Current
+        # versions of Windows use per-user temp directories with strong
+        # permissions, but I'd rather not make (poor) assumptions.
+        return super("cmd.exe /c \"\"#{native_path(command(:powershell))}\" #{legacy_args} -Command - < \"#{native_path}\"\"", check)
+      end
+    else
+      result = ps_manager.execute(command)
+
+      stdout      = result[:stdout]
+      stderr       = result[:stderr]
+      exit_code = result[:exitcode]
+
+      unless stderr.nil?
+        stderr.each do |er|
+          er.each { |e| Puppet.debug "STDERR: #{e.chop}" } unless er.empty?
+        end
+      end
+
+      Puppet.debug "STDERR: #{result[:errormessage]}" unless result[:errormessage].nil?
+
+      stuff = Puppet::Util::Execution::ProcessOutput.new(stdout.to_s || '', exit_code)
+
+      return stuff, stuff
```

Comment:
Which part is 'umm...'? The poor naming or the array return? The array return comes from mimicking the exec provider https://github.com/puppetlabs/puppet/blob/master/lib/puppet/provider/exec.rb#L74-L77